### PR TITLE
hook scrubber as logrus formatter

### PR DESCRIPTION
**Pull request by akosyakov**
## Description

We reconsidered approach and decided to redact all nested properties if they don't implement `Scrubbable` interface.

<details>
  <summary>Previous attempt</summary> I tried first to scrub before formatting and wasted too much time tuning scrubbing itself (although did fix some cases). 
  But it turned out that scrubbing mutates data as for now, so we have to make the deep copy otherwise it breaks for 
  instance startup of image build and public api. So I went for JSON formatting first and then scrubbing of JSON. It implies 
  though additional unmarshalling/marshalling. It makes logging 2-3x times slower. I am not sure we should invest more 
  time in tuning since benchmarking still shows that on average for complex data structures it takes 10000 ns.
</details>

I also explored using logurs hooks, but they don't have impact on the control flow, see [[1](https://github.com/sirupsen/logrus/blob/352781de903c9dc639752a3ac08148132746e180/entry.go#L242)]

## Related Issue(s)

fix SID-285

## How to test

- Start a workspace.
- Inspect different go component logs for some noisy errors when scrubbing prevents logging.
- Check public api logs for instance for redact.

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - ak-logrus-scrub</li>
	<li><b>🔗 URL</b> - <a href="https://ak-logrus-scrub.preview.gitpod-dev.com/workspaces" target="_blank">ak-logrus-scrub.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [x] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>
 
/hold

### CODE REVIEW

File: components/common-go/log/log_test.go
Comment (akosyakov) [inline components/common-go/log/log_test.go:113]: @csweichel @mads-hartmann it doubles formatting time because of reflection/regex/hashing for simpler cases

To be fair we don't suppose to log often? Only in error case in production?

**Comment by akosyakov** [comment]
@csweichel Is it fine to replace https://github.com/gitpod-io/gitpod/blob/main/components/common-go/log/redact.go with scrubbing or better in another PR?

File: components/common-go/log/log_test.go
Comment (akosyakov) [inline components/common-go/log/log_test.go:118]: comment: Let's for now not do it, since usually sensitive info in fields

File: components/scrubber/scrubber.go
Comment (akosyakov) [inline components/scrubber/scrubber.go:175]: (note) reflectwalk is always recursive already, there is no need to call it again

File: components/scrubber/scrubber.go
Comment (akosyakov) [inline components/scrubber/scrubber.go:193]: (note) otherwise it will double redact the property here and then in primitive callback

File: components/scrubber/scrubber.go
Comment (akosyakov) [inline components/scrubber/scrubber.go:238]: (note) ideally we should skip here as well to be sure that primitive won't double redact the value, but reflectwalk does not support skipping. Luckily testing showed that kind then is interface not string in primitive, so double redacting does not happen, and we have tests for it.

File: components/scrubber/scrubber_test.go
Comment (akosyakov) [inline components/scrubber/scrubber_test.go:95]: was not correct before, double redacted

File: components/scrubber/scrubber.go
Comment (akosyakov) [inline components/scrubber/scrubber.go:118]: it allows to cover more properties, i.e. token will cover as well `owner_token`, and `contexturl` covers such common fields like `workspaceContextURL` or `wsContextUrl` without a need to list each

File: components/scrubber/scrubber.go
Comment (akosyakov) [inline components/scrubber/scrubber.go:114]: (note) we don't expect many various fields in different cases in the same component, it allows to save allocation by avoiding conversions to lower case and boost performance to compensate for regex below

File: components/common-go/log/log.go
Comment (akosyakov) [inline components/common-go/log/log.go:98]: Maybe it should be default behaviour for `Struct` itself, i.e. traverse without reflect using type checks and only call to reflectwalk for structs. The code will be a bit more complex, since we need to switch between 2 modes, but avoiding using reflection should lead to less allocations and a bit faster processing.

Comment (csweichel) [reply components/common-go/log/log.go:98]: That makes a lot of sense. Wanna do that in a separate PR or in this one, given that you've already written the code?

File: components/scrubber/scrubber.go
Comment (csweichel) [reply components/scrubber/scrubber.go:118]: We do have a few places where we spell things like `workspaceID` and `workspaceId`. We could just list them explicitly, exp. with the regexp matcher below.

File: components/scrubber/scrubber.go
Comment (csweichel) [inline components/scrubber/scrubber.go:121]: nit: `regexp` is the name of the package as well which makes this a bit clunky. `expr` could be an alternative

File: components/scrubber/scrubber.go
Comment (csweichel) [reply components/scrubber/scrubber.go:251]: Isn't that what `return reflectwalk.SkipEntry` does?

File: components/scrubber/scrubber.go
Comment (csweichel) [inline components/scrubber/scrubber.go:211]: What requires this change compared to `MapElem`?

File: components/scrubber/scrubber.go
Comment (csweichel) [inline components/scrubber/scrubber.go:127]: How big will this map grow? How do we make sure we're not bursting at the seams here, if e.g. we feed this with an instance ID as key?

File: components/common-go/log/log.go
Comment (akosyakov) [reply components/common-go/log/log.go:98]: I tried it is hard to generalize since type checks should be specific. It did not lead to any performance improvement for logging and everything should be reimplemented twice with type checks or reflection.

File: components/scrubber/scrubber.go
Comment (akosyakov) [reply components/scrubber/scrubber.go]: I won't expect it should happen in normal usage, hope noone is putting uuids as keys. But I put a limit of 1Mb on len of all keys and added simple eviction algorithm which delete least frequent keys in batches till it gets enough memory.

File: components/scrubber/scrubber.go
Comment (csweichel) [inline components/scrubber/scrubber.go:99]: The use of `frequency` here looks a lot like an LRU. We use https://github.com/hashicorp/golang-lru for LRU caches in multiple places already. How suitable would that be here?

**Review by csweichel — COMMENTED** [review]
Given the cost of scrubbing, how much sense does it make to support selectively disabling it, i.e. on a per log-statement basis?

File: components/scrubber/scrubber.go
Comment (akosyakov) [reply components/scrubber/scrubber.go:99]: lru just kick out some? not less frequently used, I think we will need to use lfu

I tried adding some library like https://dgraph.io/blog/post/introducing-ristretto-high-perf-go-cache/ but it makes generally scrubbing slower.

I think here it is a rather safety guard, not something what is going to happen a lot with proper usage of logging to compromise performance with libraries.

File: components/scrubber/scrubber_test.go
Comment (akosyakov) [reply components/scrubber/scrubber_test.go:250]: Unfortunately it does not work because I use b.RunParallel.

**Comment by akosyakov** [comment]
> Given the cost of scrubbing, how much sense does it make to support selectively disabling it, i.e. on a per log-statement basis?

@csweichel can we instead add API to logger something like `logSafe`, `infoSafe` and so on which limits fields types, so we can scrub them before json formatting. We could deprecate then it tell yet new ones.

**Comment by akosyakov** [comment]
@csweichel @easyCZ I have 2 more questions:
- should we scrub workspace id in SaaS or for SaaS we define different set of fields?
- I've added scrubbing of nested structures, but should we add a way to opt-out, i.e. if a struct implements `Scrubbable` interface or something like that. Interface can have for instance only method like `shouldScrub: bool` ?

File: components/common-go/log/log.go
Comment (easyCZ) [inline components/common-go/log/log.go:154]: In go, the style is to have the least indentation on the longest part of the function. This means that we'd do

```
if f.skip {
	return f.JSONFormatter.Format(..)
}

for key, value := range entry.Data {
			if key == "error" || key == "severity" || key == "message" || key == "time" || key == "serviceContext" || key == "context" {
				continue
			}
			switch v := value.(type) {
			case string:
				entry.Data[key] = scrubber.Default.KeyValue(key, v)
			case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64, float32, float64, complex64, complex128:
				// no-op
			case bool:
				// no-op
			case scrubber.Scrubbable:
				err := v.Scrub(scrubber.Default)
				if err != nil {
					entry.Data[key] = "[redacted:failed]"
					fmt.Fprintf(os.Stderr, "cannot scrub log entry: %v\n", err)
				}
			default:
				entry.Data[key] = "[redacted:nested]"
			}
		}

return ...
```

File: components/image-builder-mk3/pkg/orchestrator/orchestrator.go
Comment (easyCZ) [inline components/image-builder-mk3/pkg/orchestrator/orchestrator.go:153]: Handler errors?

**Comment by easyCZ** [comment]
> Given the cost of scrubbing, how much sense does it make to support selectively disabling it, i.e. on a per log-statement basis?

> cc @easyCZ: I'd be interested in your take here as well :)

IMO, scrubbing should either be disabled, or enabled, but not selectively on lines. It makes it harder to reason about what will happen, and doesn't default to "doing the right thing"

**Comment by easyCZ** [comment]
> should we scrub workspace id in SaaS or for SaaS we define different set of fields? It was useful so far on SaaS to get from workspace id in logs to other data, i.e. analytics, traces.

SaaS should exactly match Dedicated. We shouldn't create yet another split in the config between the deployment modes

> I've added scrubbing of nested structures, but should we add a way to opt-out, i.e. if a struct implements Scrubbable > interface or something like that. Interface can have for instance only method like shouldScrub: bool ?

I'd say leave this out until we have a concrete use-case. In general, we can either choose to solve the majority of the problem centrally - e.g in the scrubber, or we can go for a more decentralized approach with the interface check. Given we've done the majority centrally, I wouldn't add the interface check until we really need it

File: components/image-builder-mk3/pkg/orchestrator/orchestrator.go
Comment (akosyakov) [reply components/image-builder-mk3/pkg/orchestrator/orchestrator.go:153]: I think all debug logs in this file can be deleted, since they are going to log `[redacted:nested]` anyway. Alternatively another data structure should be used implementing `Scrubbable` then we don't need to fiddle with marshal/unmarshall here. cc @kylos101 not sure what is the best here

Comment (akosyakov) [reply components/image-builder-mk3/pkg/orchestrator/orchestrator.go:153]: actually looking at the code it used to be the same before, so it seems errrs are skipped there on purpose

Comment (easyCZ) [reply components/image-builder-mk3/pkg/orchestrator/orchestrator.go:153]: Will `scrubber.Default.JSON` not panic when it receives `nil` as `reqs`?

Comment (akosyakov) [reply components/image-builder-mk3/pkg/orchestrator/orchestrator.go:153]: maybe, it seems it never happened 🤷‍♂️ new and old code are identical, they both will do

```
var v any
err := json.Unmarshal(nil, v)
```

json.Unmarshal return an err in this case but does not panic

i will leave it to the workspace team to decide, I think the semantic here as before.

**Comment by akosyakov** [comment]
@easyCZ @csweichel Anything else for this PR?

**Comment by akosyakov** [comment]
I moved general improvements to scrubber into https://github.com/gitpod-io/gitpod/pull/17934 to unblock https://github.com/gitpod-io/gitpod-dedicated/pull/1332

File: components/image-builder-mk3/pkg/orchestrator/orchestrator.go
Comment (akosyakov) [reply components/image-builder-mk3/pkg/orchestrator/orchestrator.go]: I've deleted them because they are scrubbed completely anyway.

File: components/ide-metrics/pkg/errorreporter/error_reporter.go
Comment (akosyakov) [inline components/ide-metrics/pkg/errorreporter/error_reporter.go:48]: @iQQBot Could you review please from IDE perspective? Do you see any other places which we scrub specially for logging or the rest is covered by common-go?

File: components/common-go/log/log.go
Comment (akosyakov) [reply components/common-go/log/log.go:154]: An idea is that everybody should use it as a feedback loop to reconsider logs. Most likely in such places nested objects should not be logged, but only some bits.

File: components/scrubber/scrubber.go
Comment (akosyakov) [reply components/scrubber/scrubber.go:62]: You mean it would be better if it returns a value? So it would be up to a client to decide whether to copy or mutate in the place?

Comment (akosyakov) [reply components/scrubber/scrubber.go:62]: It is actually complicated because how we scrub things. Scrubbing does mutate the structure in general now.

I think maybe we can use another approach similar to what VS Code does and introduce:
```golang
type TrustedValue struct {
    value any
}
```

Scrubber will trust that client code already scrubbed an underlying value in the appropriate way and use such value.

Comment (akosyakov) [reply components/scrubber/scrubber.go:62]: Talked with @csweichel that it should be rather an interface.

Comment (akosyakov) [reply components/scrubber/scrubber.go]: @iQQBot I pushed new changes, does it make more sense to you?

Comment (akosyakov) [reply components/scrubber/scrubber.go]: not necessary, it can be the same struct

**Review by easyCZ — APPROVED** [review]

**Review by easyCZ — APPROVED** [review]

**Review by mads-hartmann — APPROVED** [review]
Awesome work iterating on this and getting it through.

**Review by WVerlaek — APPROVED** [review]

**Pull request merged**
